### PR TITLE
fix free cache in MPEngine branch

### PR DESCRIPTION
**Pull request by JimyMa**
send free signal in decode engine for release cache immediately when
 migration done

### CODE REVIEW

**Comment by grimoire** [comment]
Please solve the comflicts

File: lmdeploy/pytorch/engine/mp_engine/mp_engine.py [hunk 1]
```diff
@@ -176,8 +177,8 @@ async def _mp_proc_async(server, engine):
         server.register_method('end_session', engine.end_session)
         server.register_method('get_engine_config', engine.get_engine_config)
         server.register_method('get_model_config', engine.get_model_config)
-        server.register_method('p2p_initialize', engine.p2p_initialize)
-        server.register_method('p2p_connect', engine.p2p_connect)
+        server.register_method('p2p_initialize', engine.engine_conn.p2p_initialize)
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/mp_engine/mp_engine.py]: It is not good to expose fields of engine. Any engine user should only visit the API provided by engine itself.

File: lmdeploy/pytorch/models/deepseek_v2.py [hunk 2]
```diff
@@ -724,6 +724,14 @@ def forward(self, hidden_states: torch.Tensor):
         batch_size, sequence_length, hidden_dim = hidden_states.shape
         hidden_states = hidden_states.view(-1, hidden_dim)
         topk_weights, topk_ids = self.gate(hidden_states)
+
+        if True:
```
Comment (grimoire) [inline lmdeploy/pytorch/models/deepseek_v2.py]: This is for debug?

File: lmdeploy/pytorch/paging/scheduler.py [hunk 3]
```diff
@@ -176,6 +177,23 @@ def _reorder_migrating():
             self.block_manager.allocate(seq)
             _to_running(seq)
 
+        print({
```
Comment (grimoire) [inline lmdeploy/pytorch/paging/scheduler.py]: Remove this

File: lmdeploy/pytorch/paging/scheduler.py [hunk 4]
```diff
@@ -229,6 +247,23 @@ def _reorder_waiting():
             self.block_manager.allocate(seq)
             _to_running(seq)
 
+            print({
```
Comment (grimoire) [inline lmdeploy/pytorch/paging/scheduler.py:250]: Remove this.

File: lmdeploy/pytorch/paging/scheduler.py (hunk 3 above)
Comment (grimoire) [reply lmdeploy/pytorch/paging/scheduler.py]: And all print in the scheduler.

File: requirements/serve.txt [hunk 5]
```diff
@@ -1,3 +1,4 @@
 gradio
 protobuf
+pyzmq
```
Comment (grimoire) [inline requirements/serve.txt]: pyzmq requirements has been moved to runtime.txt

File: lmdeploy/pytorch/engine/engine.py [hunk 6]
```diff
@@ -1028,6 +1037,10 @@ async def _async_loop_main(
                 await has_runable_event.wait()
                 scheduler.collect_migration_done()
                 forward_inputs, next_running = await inputs_maker.send_next_inputs()
+                if next_running is None:
+                    # self.perfill_watermark_event.wait()
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/engine.py:1074]: Better add a warning log here

File: lmdeploy/pytorch/engine/mp_engine/mp_engine.py (hunk 1 above)
Comment (JimyMa) [reply lmdeploy/pytorch/engine/mp_engine/mp_engine.py]: done in next commit

File: requirements/serve.txt (hunk 5 above)
Comment (JimyMa) [reply requirements/serve.txt]: merged latest

File: lmdeploy/pytorch/engine/engine.py (hunk 6 above)
Comment (JimyMa) [reply lmdeploy/pytorch/engine/engine.py:1074]: done in next commit.

File: lmdeploy/pytorch/paging/scheduler.py [hunk 7]
```diff
@@ -176,6 +177,23 @@ def _reorder_migrating():
             self.block_manager.allocate(seq)
             _to_running(seq)
 
+        logger.debug({
```
Comment (grimoire) [inline lmdeploy/pytorch/paging/scheduler.py]: The log would call a lot of python functions which might harm the performance.
I think we better place it in `if logger.level <= logging.DEBUG:`

Comment (JimyMa) [reply lmdeploy/pytorch/paging/scheduler.py]: shall we remove these logging because it can be achieved by metrics monitor?

File: lmdeploy/pytorch/engine/mp_engine/mp_engine.py [hunk 8]
```diff
@@ -3,14 +3,20 @@
...
 import torch.multiprocessing as mp
 
 from lmdeploy.messages import PytorchEngineConfig
+from lmdeploy.pytorch.disagg.conn.protocol import (DistServeConnectionRequest, DistServeDropConnectionRequest,
+                                                   DistServeInitRequest)
 from lmdeploy.utils import get_logger
 
 logger = get_logger('lmdeploy')
 
+if TYPE_CHECKING:
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/mp_engine/mp_engine.py:17]: Why do we need this branch?

Comment (JimyMa) [reply lmdeploy/pytorch/engine/mp_engine/mp_engine.py:17]: Here, creating a branch and using `TYPE_CHECKING` to import the Engine module avoids circular references while still enabling IDE highlighting and navigation.

**Review by grimoire — APPROVED** [review]
LGTM

**Review by grimoire — APPROVED** [review]
